designs/default-zero-value.md

```diff
@@ -0,0 +1,888 @@
+- Set: an empty set (`[]`).
+- Map: an empty map (`{}`).
+- Structure: no zero value. Structures are always nullable and never have a
+  default value. In order for a structure to have a default value, all members
+  of the structure would need to be optional. Because it is a backward
+  compatible change to remove the `@required` trait from a structure member,
+  whether a structure has a zero value or not is too volatile to rely on for
+  code generation.
+- Union: no zero value. Unions are always nullable, MUST be set to one of its
+  variants for it to be valid, and have no default variant. Adding some kind of
+  special "default" or "unset" value to unions in code to account for default
+  values would ultimately require _more_ special-case code in generated code to
+  use than just dealing with nullable unions.
+
+
+### Impact on API design
+
+Members marked with the `@default` trait cannot differentiate between whether a
+property was set to its zero value or if a property was omitted. Default zero
+values are part of the type system of Smithy and not specific to any particular
+serialization of a shape. This makes the `@default` trait a poor choice for
+Update/Patch-style APIs where it is critical to differentiate between omitted
+values and explicitly set values.
+
+To help guide API design, built-in validation will be added to Smithy to detect
+and warn when `@default` members are detected in the top-level input of
+operations that start with `Update`, operation bound to the `update` lifecycle
+of a resource, or operations that use the `PATCH` HTTP method.
+
+For example, the following model:
+
+```
+$version: "2.0"
+namespace smithy.examnple
+
+operation UpdateUser {
+    input: UpdateUserInput
+}
+
+structure UpdateUserInput {
+    @default
+    username: String
+}
+```
+
+Would emit a warning similar to the following:
+
+```
+WARNING: smithy.example#UpdateUserInput$userName (DefaultOnUpdate)
+     @ /path/to/main.smithy
+     |
+ 4   | operation UpdateUser {
+     | ^
+     = This update style operation has top-level input members marked with the
+       @default trait. It will be impossible to tell if the member was omitted
+       or explicitly provided. Affected members: [UpdateUserInput$username].
+```
+
+
+### Impact on constraint trait validation
+
+Modeled `@length`, `@pattern`, and `@range` traits need to account for zero
+values introduced by `@default` traits. Smithy will detect when a constraint
+can never be satisfied by a zero value and fail during model validation.
+
+For example, consider the following structure:
+
+```
+structure Foo {
+    @required
+    @length(min: 1, max: 100)
+    baz: String
+}
+```
+
+If this structure is later updated so that `baz` is no longer required but has
+the `@default` trait, then the following constraint trait can never hold true
+when `baz` uses the zero value:
+
+```
+structure Foo {
+    @default
+    @length(min: 1, max: 100) // <-- ERROR: default value can never satisfy min > 1
```

Comment:
I made a change here to say that constraint traits need to be aware of zero values. That is, if you say a number with a `@default` has to be between 10 and 15, then it can be `0` or any number between 10 and 15, inclusive. This will make it easier for teams to actually validate their constraints without needing to write custom code, and documentation will be more accurate about what the actual constraints are.